### PR TITLE
Refactor: remvoe `Copy` bound from `NodeId`

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

##### Refactor: remvoe `Copy` bound from `NodeId`

The `NodeId` type is currently defined as:

```rust
type NodeId: .. + Copy + .. + 'static;
```

This commit removes the `Copy` bound from `NodeId`.
This modification will allow the use of non-`Copy` types as `NodeId`,
providing greater flexibility for applications that prefer
variable-length strings or other non-`Copy` types for node
identification.

This change maintain compatibility by updating derived `Copy`
implementations with manual implementations:

```rust
// Before
#[derive(Copy...)]
pub struct LogId<NID: NodeId> {}

// After
impl<NID: Copy> Copy for LogId<NID> {}
```

- Fix: #1250

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1255)

### CODE REVIEW

File: openraft/src/raft_state/mod.rs [hunk 1]
```diff
@@ -233,17 +233,17 @@ where C: RaftTypeConfig
...
                 "new accepted.committed_vote {} must be >= current accepted.committed_vote: {}",
                 new_vote,
                 current_vote.display(),
             );
         }
 
-        if Some(accepted) > curr_accepted {
-            self.io_state.io_progress.accept(accepted);
+        if Some(accepted.clone()) > curr_accepted {
+            self.io_state.io_progress.accept(accepted.clone());
```
Comment (SteveLauC) [inline openraft/src/raft_state/mod.rs:246]: Let's do this:

```suggestion
        if Some(&accepted) > curr_accepted.as_ref() {
            self.io_state.io_progress.accept(accepted);
```

File: openraft/src/replication/mod.rs [hunk 2]
```diff
@@ -391,14 +391,14 @@ where
...
-                let r = LogIdRange::new(rng.prev, rng.prev);
+                let r = LogIdRange::new(rng.prev.clone(), rng.prev.clone());
                 (vec![], r)
             } else {
                 // limited_get_log_entries will return logs smaller than the range [start, end).
                 let logs = self.log_reader.limited_get_log_entries(start, end).await?;
 
-                let first = *logs.first().map(|x| x.get_log_id()).unwrap();
-                let last = *logs.last().map(|x| x.get_log_id()).unwrap();
+                let first = logs.first().map(|x| x.get_log_id().clone()).unwrap();
```
Comment (SteveLauC) [inline openraft/src/replication/mod.rs:400]: Can be avoided:

```suggestion
                let first = logs.first().map(|x| x.get_log_id()).unwrap();
```

File: openraft/src/replication/mod.rs [hunk 3]
```diff
@@ -496,7 +496,7 @@ where
                 }))
             }
             AppendEntriesResponse::Conflict => {
-                let conflict = sending_range.prev;
+                let conflict = sending_range.prev.clone();
```
Comment (SteveLauC) [inline openraft/src/replication/mod.rs:499]: Looks like we can move `sending_range.prev` here

**Review by SteveLauC — COMMENTED** [review]
Sorry for reviewing this PR on GitHub, reviewable does not show line numbers, which makes me impossible to locate the code and do a check in my IDE, left some comments.

BTW, there are multiple "unused" warnings reported by GitHub Actions on the review page, are they false-positive? If not, perhaps we can do a clean-up:)

**Review by schreter — COMMENTED** [review]
@SteveLauC (no review from my side, just a tip)

> reviewable does not show line numbers, which makes me impossible to locate the code and do a check in my IDE, left some comments

Actually, it does. It is just a bit cumbersome, since you have to set it up, [see here](https://docs.reviewable.io/tips#line-numbers). Similarly, for locating the code in the IDE, you can directly jump into the IDE from Reviewable from any comment (also from a new draft). It is documented [here](https://docs.reviewable.io/accountsettings#external-editor-line-link-template). With that, you don't even strictly need line numbers (BTW, those are shown in the draft of the new comment as well), but I personally did configure them for myself :-).

> *[Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1255#-:-O95Vony1sJUSwMHAHo9:b-y85d8l)* status: 0 of 57 files reviewed, 4 unresolved discussions (waiting on @drmingdrmer and @SteveLauC)

**Comment by SteveLauC** [comment]
> @SteveLauC (no review from my side, just a tip)

Amazing! TIL! Will set them up. 😄

**Review by SteveLauC — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1255)* status: 0 of 57 files reviewed, 4 unresolved discussions (waiting on @drmingdrmer)

**Comment by drmingdrmer** [comment]
@SteveLauC 
Thank you for addressing these details. Your attention to these nuances is greatly appreciated.

@schreter 
The Reviewable hints are indeed very helpful!

Unfortunately, I haven't figured out how to open CLion directly from Reviewable yet. My browser can open VS Code using the `vscode://` protocol, but it doesn't work with `clion://` for CLion. 

I've learned that I need to manually register a URL handler for CLion. I'll try to set this up later. It's a bit disappointing that it's not as straightforward as with VS Code. :(

**Pull request merged**